### PR TITLE
[Google Blockly] Support mini-toolboxes in Google Blockly labs

### PR DESCRIPTION
**Pull request by mikeharv**
## Overview

This PR implements the fields `CdoFieldToggle` and `CdoFieldFlyout` and the new horizontal flyout `CdoBlockFlyout` to add support for what we colloquially refer to as a "mini-toolbox". The toggle field provides the means to create a flyout in a field on the block which contains other blocks meant to be used in conjunction with it.
|CDO Blockly|Google Blockly (new)|
|-|-|
|cdo|gb|

_Note:_ In CDO Blockly labs, the blocks that we typically surface in a mini-toolbox also have a special image "shadowing" feature that replaces the word "sprite" with a costume thumbnail. That feature is out of scope for this task.

This PR builds off of the following work:
- https://github.com/code-dot-org/code-dot-org/pull/52715
- https://github.com/code-dot-org/code-dot-org/pull/52708
- https://github.com/code-dot-org/code-dot-org/pull/52711

Started by @bencodeorg and continued by @fisher-alice and myself with major help from @BeksOmega from Blockly.

## Technical Approach
Most Sprite Lab blocks today are defined on levelbuilder as config files that are passed through a complicated flow of operations found in `block_utils.js`: https://github.com/code-dot-org/code-dot-org/blob/870f7d59bdec5d73f4bf9748b94a0cfcd07e4ac8/apps/src/block_utils.js#L860-L1277

A simplified block config looks like this:
```
{
  "func": "spriteClicked",
  "blockText": "when {SPRITE} clicked",
  "miniToolboxBlocks": [
    "gamelab_clickedSpritePointer"
  ],
  "args": [
    {
      "name": "SPRITE",
      "type": "Sprite"
    }
  ]
}
```

In cases where a block a certain kind of input (including field), an object is added to the `args` list. The `block_utils.js` flow interprets each of these and then determine which inputs and fields are actually needed. Mini-toolbox blocks are provided separately as a simple list of block types. When this list is present, and when `appOptions` dictate that the level should have them, a mini-toolbox is created on the block. The way in which we make the field(s) and it's associated flyout varies greatly between CDO and Google Blockly implementations, so the logic has been pulled out to helper files. The new logic we are concerned with has been added to the `spriteLabBlock.js` helper file.

In CDO Blockly, we were able to create all of the needed block elements at once and then carry on with creating the other fields and inputs. This was possible due to extensive customizations in the Blockly fork that are not present in mainline. Our new implementation works as follows:
1. `initializeMiniToolbox` - Create a toggle field, but don't actually add it to an input yet.
  - The toggle field knows how to create a flyout field.
  - The toggle field can be clicked to show/hide the flyout field.
  - The toggle is created with +/- icons and custom blue/white style overrides.
2. Back in `block_utils`, the other inputs that were determined by the config `args` are created and added to the block.
3. `appendMiniToolboxToggle` - Finish setting up the block's mini-toolbox
  - The toggle field created in step 1 is added to the block at the beginning of the first input row.
  - A callback is registered using the list of miniToolboxBlocks - this tells Blockly which blocks to create in the flyout
  - Serialization hooks are added to handle the extra state. We need to save whether the flyout is open or closed in order to make it possible for levels to have these blocks show with the flyout open-by-default. The JSON hooks are `saveExtraState` and `loadExtraState` (used frequently by Music Lab), and the XML hooks are `mutationToDom` and `domToMutation`. We serialize the current icon shown on the toggle field. When deserializing, we match the icon to what we have saved, and if it's not the default icon (ie. it's showing `-`), we create the flyout. 

There was a lot of manual bug bashing to get these blocks working as expected. One nice bonus is that the RTL versions of these blocks are actually a bit janky on CDO/production and have been greatly improved in the new implementation.

|CDO Blockly|Google Blockly (new)|
|-|-|
|image|image|

Another nice bonus is that these blocks now work in Poetry lab for the first time!

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js [hunk 1]
```diff
@@ -1,5 +1,102 @@
+import _ from 'lodash';
+import {SVG_NS} from '@cdo/apps/constants';
+import Button from '@cdo/apps/templates/Button';
+
 // This file contains customizations to Google Blockly Sprite Lab blocks.
 
+const dummyInputType = 5;
```
Comment (molly-moen) [inline apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js]: Can you add a comment explaining what this field means?

File: apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js [hunk 2]
```diff
@@ -1,5 +1,102 @@
...
+      const flyoutField = new Blockly.FieldFlyout(_, {
+        flyoutKey: `flyout_${block.type}`,
+        sizingBehavior: 'fitContent',
+        name: 'FLYOUT',
+      });
+      block
+        .appendDummyInput('flyout_input')
+        .appendField(flyoutField, `flyout_${block.type}`);
+      return flyoutField;
+    };
```
Comment (molly-moen) [inline apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js:22]: super-nit: add a blank line after each of these inline functions to make things more readable?

File: apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js [hunk 3]
```diff
@@ -1,5 +1,102 @@
...
+    return flyoutToggleButton;
+  },
+  appendMiniToolboxToggle(miniToolboxBlocks, flyoutToggleButton) {
+    this.setInputsInline(true);
+    const firstInput = this.inputList[0];
+
+    firstInput.insertFieldAt(0, flyoutToggleButton, `button_${this.type}`);
+    const lastInput = this.inputList[this.inputList.length - 1];
+    if (lastInput.type !== dummyInputType) {
+      this.appendDummyInput();
```
Comment (molly-moen) [inline apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js:73]: why is this needed?

File: apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js [hunk 4]
```diff
@@ -1,5 +1,102 @@
...
+        useDefaultIcon: flyoutToggleButton.useDefaultIcon,
+      };
+    };
+
+    this.loadExtraState = function (state) {
+      const useDefaultIcon = state['useDefaultIcon'];
+      flyoutToggleButton.setIcon(useDefaultIcon);
+    };
+
+    this.mutationToDom = function () {
```
Comment (molly-moen) [inline apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js:106]: mind adding a comment explaining what a mutation is?

**Review by molly-moen — APPROVED** [review]
Still learning here, so just some clarifying questions. The code changes seem reasonable overall!

File: apps/src/block_utils.js [hunk 5]
```diff
@@ -1121,6 +1123,19 @@ exports.createJsWrapperBlockCreator = function (
         }
         interpolateInputs(blockly, this, inputRows, inputTypes, inline);
         this.setInputsInline(inline);
+
+        // Use window.appOptions, not global appOptions, because the levelbuilder
+        // block page doesn't have appOptions, but we *do* want to show the mini-toolbox
+        // there
+        if (
+          miniToolboxBlocks &&
```
Comment (fisher-alice) [inline apps/src/block_utils.js]: You could use `showMiniToolbox` constant already declared and assigned above instead of condition/comment.

**Review by fisher-alice — APPROVED** [review]
Woo hoo! Just a minor suggestion. 🎉

File: apps/src/block_utils.js (hunk 5 above)
Comment (mikeharv) [reply apps/src/block_utils.js]: I actually have this change made locally but had forgotten to push it! Thanks, Alice!

File: apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js (hunk 3 above)
Comment (mikeharv) [reply apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js:73]: Just added a comment. These blocks require this plugin (https://github.com/google/blockly-samples/tree/master/plugins/renderer-inline-row-separators) which modifies our renderer so that dummy inputs are treated as row separators. We need to make sure that the top row of the block ends in a row separator or else the flyout field will be rendered to the right of the other fields, instead of below.

File: apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js (hunk 4 above)
Comment (mikeharv) [reply apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js:106]: Added a general explanation of mutators above `saveExtraState`, additional comments for each pair of hooks, and examples of how the extra state would be saved.

**Comment by mikeharv** [comment]
@breville Heads up that I've add a commit here: https://github.com/code-dot-org/code-dot-org/pull/52842/commits/649b860d5a83dcfe281a3dab928c99b3678623f7
This will fix the regression you saw in the Music Lab block by ensuring Music Lab continues to use the base Thrasos renderer (without inline row separators). None of our other labs have blocks that are constructed quite this way so we should be good to keep using the modified renderer as our default for other labs. I've double-checked all blocks that show up in Poetry, Dance, Flappy, Bounce, and Sprite Lab.

Context for anyone else who is curious: https://codedotorg.slack.com/archives/C04TH8400AU/p1689641875315419

File: apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js (hunk 1 above)
Comment (mikeharv) [reply apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js]: Glad you asked, as I meant to clean this up!

These types are defined here: https://github.com/google/blockly/blob/master/core/inputs/input_types.ts

We can actually add this to the wrapper and access it directly instead, which feels a whole lot less fragile:

https://github.com/code-dot-org/code-dot-org/pull/52842/commits/741f10c3eaa6f79542d4ddd231c69a9a36ead506

**Comment by fisher-alice** [comment]
> @breville Heads up that I've add a commit here: [649b860](https://github.com/code-dot-org/code-dot-org/commit/649b860d5a83dcfe281a3dab928c99b3678623f7) This will fix the regression you saw in the Music Lab block by ensuring Music Lab continues to use the base Thrasos renderer (without inline row separators). None of our other labs have blocks that are constructed quite this way so we should be good to keep using the modified renderer as our default for other labs. I've double-checked all blocks that show up in Poetry, Dance, Flappy, Bounce, and Sprite Lab.
> 
> Context for anyone else who is curious: https://codedotorg.slack.com/archives/C04TH8400AU/p1689641875315419

Nice fix!

**Comment by bencodeorg** [comment]
@mikeharv I remember you saying something to me about the ordering of new code in block_utils being important, something like "`initializeMiniToolbox` must be called before x" (I can't recall what x is...) and "`appendMiniToolboxToggle` must be called after y" (I also can't recall what y is, and y might be the same as x). Am I remembering that correctly? If so, is there any way we could make that ordering more explicitly understood (a comment?)

**Comment by mikeharv** [comment]
@bencodeorg `initializeMiniToolbox` returns the field that is needed by `appendMiniToolboxToggle` which is why they need to be called in that order. Moving all of the logic out of `block_utils.js` was more about trying to streamline that file than it was about making utilities that we'd necessarily reuse.

> @mikeharv I remember you saying something to me about the ordering of new code in block_utils being important, something like "`initializeMiniToolbox` must be called before x" (I can't recall what x is...) and "`appendMiniToolboxToggle` must be called after y" (I also can't recall what y is, and y might be the same as x). Am I remembering that correctly? If so, is there any way we could make that ordering more explicitly understood (a comment?)

**Comment by bencodeorg** [comment]
> _@bencodeorg `initializeMiniToolbox` returns the field that is needed by `appendMiniToolboxToggle` which is why they need to be called in that order. Moving all of the logic out of `block_utils.js` was more about trying to streamline that file than it was about making utilities that we'd necessarily reuse._

Sorry I wasn't asking about whether these are reusable utilities, but rather if there's some new specific ordering that isn't totally obvious within block_utils that we might want to call out? I could imagine someone looking at this in a couple years and asking...why isn't the code in these two `if showMiniToolbox` statements together? Was it something about needing `interpolateInputs` to run between the two?

**Comment by mikeharv** [comment]
@bencodeorg Got it! And yes, you're correct about `interpolateInputs`. We want to add the toggle field to the start of the first input; doing so requires that the other block inputs get added first so we have somewhere to put it. 

> Sorry I wasn't asking about whether these are reusable utilities, but rather if there's some new specific ordering that isn't totally obvious within block_utils that we might want to call out? I could imagine someone looking at this in a couple years and asking...why isn't the code in these two `if showMiniToolbox` statements together? Was it something about needing `interpolateInputs` to run between the two?
